File: cmd/problem_2a/problem_2a.go
package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
	"strings"
)

// Graph struct will represent all edges.
// Key in the map       -> vertex value.
// Val for each vertex  -> list of edges.
type Graph struct {
	edges map[int][]*Edge
}

// to       -> destination vertex, where this edge goes to.
// capacity -> weight of the edge, capacity that can flow through it.
// flow     -> what is currently flowing through the edge.
type Edge struct {
	to       int
	capacity int
	flow     int
}

// NewGraph() just creates a new graph instance.
// The map will hold vertex values as the keys and their outgoing edges in a slice as the val.
// NOTE: the edges slice for each vertex will also contain any backwards edges.
func NewGraph(vertexCount int) *Graph {
	graph := &Graph{
		edges: make(map[int][]*Edge),
	}

	for i := range vertexCount {
		graph.edges[i] = []*Edge{}
	}

	return graph
}

func (graph *Graph) AddEdge(v1, v2, weight int) {
	// Forward edge (v1 -> v2).
	forward := &Edge{
		to:       v2,
		capacity: weight,
		flow:     0,
	}

	// Backward edge (v2 -> v1).
	backward := &Edge{
		to:       v1,
		capacity: 0,
		flow:     0,
	}

	// Append to respective vertex Edge slice.
	graph.edges[v1] = append(graph.edges[v1], forward)
	graph.edges[v2] = append(graph.edges[v2], backward)
}

func bronKerbosch() {}

func readFile(fileName string) (*Graph, int, int) {
	file, err := os.Open(fileName)
	if err != nil {
		log.Fatal("Error opening the file")
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)
	if !scanner.Scan() {
		log.Fatal("readFile(): Issue scanning the file -- may be empty..")
	}

	// First row (2 values) -> vertex count and then the arc count.
	firstLine := strings.Fields(scanner.Text())
	if len(firstLine) < 2 || len(firstLine) > 2 {
		log.Fatal(
			"readFile(): Error parsing the first line -- either too few arguments or too many. Should only include vertex count and arc count",
		)
	}

	vertexCount, err := strconv.Atoi(firstLine[0])
	if err != nil {
		log.Fatal("readFile(): Error reading in the vertexCount")
	}

	arcCount, err := strconv.Atoi(firstLine[1])
	if err != nil {
		log.Fatal("readFile(): Error reading in the arcCount")
	}

	graph := NewGraph(vertexCount)
	source := 0
	sink := 0
	arcReadCount := 1

	// Read in the remaining rows containing the edges (v1, v2, arc weight).
	for scanner.Scan() {

		// If there's any extra rows that are not included in the arcCount defined, break.
		if arcReadCount > arcCount {
			break
		}

		line := strings.Fields(scanner.Text())
		if len(line) < 3 || len(line) > 3 {
			log.Fatal("readFile(): vertex and weight row either has too few or too many arguments")
		}

		v1, err := strconv.Atoi(line[0])
		if err != nil {
			log.Fatal("readFile(): Error reading in v1")
		}

		v2, err := strconv.Atoi(line[1])
		if err != nil {
			log.Fatal("readFile(): Error reading in the v2")
		}

		weight, err := strconv.Atoi(line[2])
		if err != nil {
			log.Fatal("readFile(): Error reading in the weight")
		}

		// Track the sink (max v2 found).
		if v2 > sink {
			sink = v2
		}

		// Create and add edge to the graph (both forward and backward)
		graph.AddEdge(v1, v2, weight)
		arcReadCount++
	}

	return graph, source, sink
}

func main() {
	if len(os.Args) != 2 {
		log.Fatal("usage: ./bin/problem2a filename")
	}

  // build the graph
  graph, source, sink := readFile(os.Args[1])

  log.Println(graph)
  log.Println(source)
  log.Println(sink)
}

File: problem_2/src/problem_2a.py
# Zachary Perry
# COSC 581 Assignment 7: Part 2a
# 4/10/25

import sys
import time


def read_file_build_graph(file_name: str):
    '''
        The read file and build graph function does exactly that. It reads in
        the file containing the graph input and creates the graph.

        The input file cointains a graph in modified DIMACS format, where the
        first line has the total number of vertices and edges and the rest of
        the lines have two values, both vertices representing an edge between
        them.

        Input:
            file_name (str): name of the input file
        Returns:
            graph (2d arr): 2d array representing the graph as an adjacency lst
            vertices (int): total number of vertices read
    '''
    graph = []

    with open(file_name, "r") as file:
        contents = file.readlines()

    # Get the first line containing the number of vertices and edges
    vert_edge_line = contents[0].strip().split()
    if len(vert_edge_line) < 2 or len(vert_edge_line) > 2:
        print("error (read_file_build_graph): error parsing first line in the file. Should only include the vertex count and edge count")
        sys.exit(1)
    vertices = int(vert_edge_line[0])
    edges = int(vert_edge_line[1])

    # Initialize graph (adjacency list).
    for i in range(vertices):
        graph.append([])

    edges_read = 0

    # Build adjacency list.
    for i in range(1, len(contents)):
        # NOTE: Ensures we don't read extra edges
        if edges_read > edges:
            break
        curr_edge = contents[i].strip().split()
        if len(curr_edge) > 2 or len(curr_edge) < 2:
            print("error (read_file_build_graph): error parsing edge. Either too little or too few arguments.")
            sys.exit(1)

        vertex_1 = int(curr_edge[0])
        vertex_2 = int(curr_edge[1])

        graph[vertex_1].append(vertex_2)
        graph[vertex_2].append(vertex_1)

        edges_read += 1

    return graph, vertices


def bron_kerbosch(R, P, X, graph, max_clique):
    '''
         Implementation of the standard Bron Kerbosch Algorithm for
         finding the max clique in a graph.

         Source: https://en.wikipedia.org/wiki/Bron–Kerbosch_algorithm

         Input:
            R: The set of vertices currently a part of the potential clique being built
            P: The set of vertices that are all are connected to every vertex in R and could potentially be fully added
            X: The set of vertices that have already been processed
            graph (2d arr): 2d array representing the graph as an adjacency lst
            max_clique: Set containing the current vertices within the max clique found
         Returns:

    '''

    # If both P and X are empty: then report R as the current max clique.
    if not P and not X:
        if len(R) > len(max_clique[0]):
            max_clique[0] = R.copy()
        return

    # Otherwise, recursively explore whether each vertex in P can be added to the current clique.
    # 1. grab all neighbors of the current vertex
    # 2. Recursively call BK with:
        # 2a: R after adding the current vertex to the clique
        # 2b: P now equals the intersection of vertices with the current neighbors of the current vertex
        # 2c: X now equals the intersection of vertices with the current neighbors of the current vertex
    for vertex in list(P):
        curr_neighbors = set(graph[vertex])

        bron_kerbosch(
            R.union({vertex}),
            P.intersection(curr_neighbors),
            X.intersection(curr_neighbors),
            graph,
            max_clique
        )

        # Mark the current vertex as processed
        P.remove(vertex)
        X.add(vertex)


def main():

    if len(sys.argv) < 2:
        print("usage: python3 problem_2a.py filename")
        sys.exit(1)

    file_name = sys.argv[1]

    # Read in file, get graph, etc.
    graph, num_vertices = read_file_build_graph(file_name)

    start = time.time()

    # Initialize the max_clique set, R, P, and X.
    # Then, run bron_kerbosch to find the max clique.
    max_clique = [set()]
    R = set()
    P = set(range(num_vertices))
    X = set()
    bron_kerbosch(R, P, X, graph, max_clique)

    end = time.time()
    runtime = end - start

    print("Max clique: ", max_clique[0])
    print("Total time to find max clique: ", runtime, "sec")


if __name__ == "__main__":
    main()
